File: packages/bpk-component-badge/src/BpkBadge.js
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */

import PropTypes from 'prop-types';
import React from 'react';
import { cssModules } from 'bpk-react-utils';

import STYLES from './bpk-badge.css';

export const BADGE_TYPES = {
  warning: 'warning',
  success: 'success',
  destructive: 'destructive',
  light: 'light',
  inverse: 'inverse',
  outline: 'outline',
};

const getClassName = cssModules(STYLES);

const badgeTypeClassNames = {
  [BADGE_TYPES.warning]: null,
  [BADGE_TYPES.success]: getClassName('bpk-badge--success'),
  [BADGE_TYPES.destructive]: getClassName('bpk-badge--destructive'),
  [BADGE_TYPES.light]: getClassName('bpk-badge--light'),
  [BADGE_TYPES.inverse]: getClassName('bpk-badge--inverse'),
  [BADGE_TYPES.outline]: getClassName('bpk-badge--outline'),
};

export type Props = {
  type: $Keys<typeof BADGE_TYPES>,
  docked: ?string,
  centered: boolean,
  className: ?string,
};

const BpkBadge = (props: Props) => {
  const { type, docked, centered, className, ...rest } = props;
  const classNames = [getClassName('bpk-badge'), badgeTypeClassNames[type]];

  if (docked === 'right') {
    classNames.push(getClassName('bpk-badge--docked-right'));
  }
  if (docked === 'left') {
    classNames.push(getClassName('bpk-badge--docked-left'));
  }
  if (centered) {
    classNames.push(getClassName('bpk-badge--centered'));
  }
  if (className) {
    classNames.push(className);
  }

  return <span className={classNames.join(' ')} {...rest} />;
};

BpkBadge.propTypes = {
  type: PropTypes.oneOf(Object.keys(BADGE_TYPES)),
  docked: PropTypes.oneOf(['right', 'left', null]),
  centered: PropTypes.bool,
  className: PropTypes.string,
};

BpkBadge.defaultProps = {
  type: BADGE_TYPES.warning,
  docked: null,
  centered: false,
  className: null,
};

export default BpkBadge;

File: native/packages/react-native-backpack/libraries/Badge/ios/native/classes/BPKBadge.m
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#import "BPKBadge.h"

#import <Backpack/Spacing.h>
#import <Backpack/Color.h>
#import <Backpack/Radii.h>
#import <Backpack/Font.h>

NS_ASSUME_NONNULL_BEGIN
@interface BPKBadge()
@property(nonatomic, strong) UILabel *label;

- (void)setup;
@end


@implementation BPKBadge

- (instancetype)initWithType:(BPKBadgeType)type message:(NSString *)message {
    self = [self initWithFrame:CGRectZero];

    if (self) {
        self.type = type;
        self.message = message;
    }

    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];

    if (self) {
        [self setup];
    }

    return self;
}

- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];

    if (self) {
        [self setup];
    }

    return self;
}

- (void)setType:(BPKBadgeType)type {
    _type = type;

    self.layer.borderWidth = 0.0;
    switch (type) {
        case BPKBadgeTypeSuccess:
            self.backgroundColor = [BPKColor green500];
            break;
        case BPKBadgeTypeWarning:
            self.backgroundColor = [BPKColor yellow500];
            break;
        case BPKBadgeTypeDestructive:
            self.backgroundColor = [BPKColor red500];
            break;
        case BPKBadgeTypeLight:
            self.backgroundColor = [BPKColor gray50];
            break;
        case BPKBadgeTypeInverse:
            self.backgroundColor = [BPKColor white];
            break;
        case BPKBadgeTypeOutline:
            self.backgroundColor = [[BPKColor white] colorWithAlphaComponent:0.2];
            self.layer.borderColor = [BPKColor white].CGColor;
            self.layer.borderWidth = 1.0;
            break;
        default:
            break;
    }

    if (type == BPKBadgeTypeOutline || type == BPKBadgeTypeDestructive) {
        self.label.textColor = [BPKColor white];
    } else {
        self.label.textColor = [BPKColor gray700];
    }
}

- (void)setMessage:(NSString *)message {
    self.label.text = message;
}

- (NSString *)message {
    return self.label.text;
}

#pragma mark - Private

- (void)setup {
    UILabel *label = [[UILabel alloc] initWithFrame:CGRectZero];
    label.font = [BPKFont textXs];
    [self addSubview:label];
    label.translatesAutoresizingMaskIntoConstraints = NO;

    [label.leadingAnchor constraintEqualToAnchor:self.leadingAnchor constant:BPKSpacingMd].active = YES;
    [self.trailingAnchor constraintEqualToAnchor:label.trailingAnchor constant:BPKSpacingMd].active = YES;
    [label.topAnchor constraintEqualToAnchor:self.topAnchor constant:BPKSpacingSm].active = YES;
    [self.bottomAnchor constraintEqualToAnchor:label.bottomAnchor constant:BPKSpacingSm].active = YES;
    self.label = label;

    self.type = BPKBadgeTypeSuccess;
    self.layer.cornerRadius = BPKBorderRadiusSm;
    self.layer.masksToBounds = YES;
}

@end
NS_ASSUME_NONNULL_END
